packages/labs/task/src/task.ts

```diff
@@ -400,16 +423,93 @@ export class Task<
         throw new Error(`Unexpected status: ${this.status}`);
     }
   }
-
-  private _argsDirty(args?: T) {
-    const prev = this._previousArgs;
-    this._previousArgs = args;
-    return Array.isArray(args) && Array.isArray(prev)
-      ? args.length === prev.length && args.some((v, i) => notEqual(v, prev[i]))
-      : args !== prev;
-  }
 }
 
 type MaybeReturnType<F> = F extends (...args: unknown[]) => infer R
   ? R
   : undefined;
+
+export const shallowArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => !notEqual(v, newArgs[i])));
+
+export const deepArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => deepEquals(v, newArgs[i])));
+
+const objectValueOf = Object.prototype.valueOf;
+const objectToString = Object.prototype.toString;
+const {keys: objectKeys} = Object;
+const {isArray} = Array;
+
+export const deepEquals = (a: unknown, b: unknown): boolean => {
+  if (Object.is(a, b)) {
+    return true;
```

Comment:
Do we want NaN to match NaN?

Reply:
I think so. I also think matching some built-in semantics like Object.is is good. That's why I inverted the -0/0 check.